### DISCUSSION TITLE
Don't use "have" for action method names

### DISCUSSION THREAD

**Discussion by ThomasLandauer** [Codeception/Codeception#6205] [Ideas]
Follow up of https://github.com/Codeception/Codeception/issues/4257

I'm suggesting to completely drop the term "have" from method names, like e.g. `haveHttpHeader`.

Reasons:

DavertMik says:

> But for have it's totally fine as it describes what you have in the beginning.

I say: **No, no, no!** Ask ten new users what they *think* that `$I->haveUser('davert');` means.

I bet that 10 of them will answer. "Codeception will check if I have that user". And not one will answer what the function actually does, namely: "Codeception will create that user."

The common name for methods like that is **set**. Is there a specific reason why you didn't choose that? `*set*HttpHeader` certainly conveys the meaning of the function (namely to set an HTTP header :-) better than `*have*HttpHeader`! And `setProduct` or `setInDatabase` also sound good to me.

`expect` is even worse, but I can't make a better suggestion yet (since I don't know what it's being used for). Is there list of all these comments anywhere? (See also https://github.com/Codeception/Codeception/issues/4238#issuecomment-304539695 )

**Comment by acorncom** [comment]
Yes, I'd agree with this.  `set` or `add` would seem to be a clearer explanation of what is *actually* happening.  `haveHttpHeader` in my mind would be to check that the http header exists ...

**Comment by andrewandante** [comment]
Having just spent a long time looking for `addHeader()` to counterpart `deleteHeader()`, I can only lend my support to removing the leading `have`. Or at least update the corresponding methods to `dontHave`

**Comment by SamMousa** [comment]
Note that `set` and `add` are the same when it comes to scalar values, but not for arrays.
For headers for example, which can be multivalued, `add` usually implies adding a new value to the list whereas `set` implies replacing all existing values with the new given value.

With regards to `expect` we could switch to `assert` since that's what we're already doing. So instead of `expectException` we have `assertExceptionIsThrown`.
Then there's the comment part of `expect` where it has no functional use but is only used for documenting the test.
`$I->expectTo('be greeted with a nice welcome message');`
Here we have several functions:
```php
expectTo($prediction)
expect($prediction)
amGoingTo($argumentation)
am($role)
lookForwardTo($achieveValue)
```
All of these just add comments with a different prefix. These are defined [here](https://github.com/Codeception/Codeception/blob/2.5/src/Codeception/Lib/Actor/Shared/Comment.php).


The best way forward in my opinion is to create a document describing the naming rules (I know for example that the Yii module has `amLoggedInAs`), for each case.
Then we mark all invalid functions as deprecated and add the new functions.
In the next BC breaking release we can then remove all deprecated functions.

Currently I see several kinds of functions:
- Make assertion
- Add comment
- Change state

Sample rules / requirements:
- A functions' type should be easily identifiable from its name.
- State changing functions must start with `set` or `add`.
- State changing functions starting with `add` must operate array types.
- If a function starting with `add` exists, a corresponding `set` must exist as well.
- Assertion functions always start with `assert` unless they assert something about the latest response, in which case they start with `see` or `dontSee`.

And so on..

If we agree on the rules we can add them to the docs and basically with each PR we can confirm that it does not introduce new functions breaking the rules.

@DavertMik what do you think? This will offer a safe upgrade path to a future where naming is intuitively clear for everyone without requiring us to do it in one go.
